File: javascript/ValidAnagram.js
class Solution {
    /**
     * @param {string} s
     * @param {string} t
     * @return {boolean}
     */
    isValidAnagram(s, t) {

        // constraint: s and t consist of lowercase English letters.
        
        if (s.length !== t.length) return false;

        const charS = {};
        const charT = {};

        let sc, tc;

        // counting up
        for (let i = 0; i < s.length; i++) {
            sc = s[i];
            tc = t[i];

            charS[sc] = (charS[sc] || 0) + 1;
            charT[tc] = (charT[tc] || 0) + 1;
        }
        
        // check if each counter had the same count since JS === compares pointers
        for (const c in charS) {
            if (charS[c] !== charT[c]) {
                return false;
            }
        }
        return true;
    }
}

module.exports = Solution;

File: cpp/solutions/valid_anagram.cpp
#include "valid_anagram.h"
#include <unordered_map>

bool ValidAnagram::isAnagram(std::string s, std::string t) {
    if (s.length() != t.length()) {
        return false;
    }
    std::unordered_map<char, int> charS;
    std::unordered_map<char, int> charT;
    
    for (size_t i = 0; i < s.size(); i++) {
        charS[s[i]]++;
        charT[t[i]]++;
    }

    return charS == charT; // since cpp std compares object equality
};
